RFC-0009-dataloader.md

```diff
@@ -0,0 +1,410 @@
+# DataLoader architecture updates and TarDataset implementation
+# Problem statement
+This proposal aims to construct a modular, user-friendly, and performant toolset to address the ambiguous activity referred to as “dataloading” within PyTorch, a simplification attributable to the indivisibility of the DataLoader abstraction prescribed today. In reality, “dataloading” is a diverse set of operations that should be supported by extensible building blocks, out of which the present abstractions, and far more could be easily built. Some typical needs which are scarcely supported in the present implementation include:
+
+- Lazy loading - Users want to point PyTorch to a remote data source (e.g. HTTP, S3, GCP, Azure, Manifold, Hive) and iterate over the contents without downloading the entire dataset, ideally only downloading samples as-soon-needed. Further, If a user writes such code for a remote storage platform, there is no natural place to contribute it for public or private reuse.
```

Comment:
I thought I'd point out [fsspec](https://filesystem-spec.readthedocs.io/en/latest/) which provides interfaces for many of these formats, and the related [s3fs](https://fs-s3fs.readthedocs.io/en/latest/) and [gcsfs](https://gcsfs.readthedocs.io/en/latest/) built on top of it. That would be the natural place to contribute, and write a lazy Dataset abstraction on top of.